File: 6주차/박수교/뒤에 있는 큰 수 찾기.py
2) 스택 사용 풀이
def solution(numbers):
    answer = [-1] * len(numbers)
    stack = []

    for idx, num in enumerate(numbers):
        while stack and numbers[stack[-1]] < num:
            answer[stack.pop()] = num
            # print("a")
            # print(stack)
            # print(answer)
        stack.append(idx)
        # print("b")
        # print(stack)
        # print(answer)
    return answer

1) 첫번째 풀이 -> 시간초과(20 - 23)
def solution(numbers):
    N = len(numbers)
    flag = 0
    for i in range(N):
        for j in range(i+1, N):
            if numbers[i] < numbers[j]:
                flag = -1
                numbers[i] = numbers[j]
                break
        if flag == -1:
            flag = 0
            continue
        else:
            numbers[i] = -1
    return numbers

File: 6주차/김용상/뒤에 있는 큰 수 찾기.java
import java.util.*;
class Solution {
    public int[] solution(int[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        Arrays.fill(answer,-1);
        // n log n으로 끝내야함
        Stack<Node>st = new Stack<Node>();
        for (int i=0;i<n;i++){
            while(!st.isEmpty()){
                Node peek = st.peek();
                if (numbers[i]>peek.num){
                    st.pop();
                    answer[peek.idx] = numbers[i];
                    continue;
                }
                break;
            }                  
            st.push(new Node(numbers[i],i));
        }
        return answer;
    }
}
class Node{
    int num;
    int idx;
    public Node(int n,int x){
        this.num = n;
        this.idx = x;
    }
} 
